src/libasr/pass/unique_symbols.cpp

```diff
@@ -0,0 +1,602 @@
+                ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+                sym_to_renamed[sym] = update_name(x.m_name);
+            }
+        }
+        for (auto &a : x.m_symtab->get_scope()) {
+            this->visit_symbol(*a.second);
+        }
+    }
+
+    void visit_Variable(const ASR::Variable_t &x) {
+        if (all_symbols_mangling || should_mangle) {
+            ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+            sym_to_renamed[sym] = update_name(x.m_name);
+        }
+    }
+
+    void visit_ClassType(const ASR::ClassType_t &x) {
+        if (x.m_abi != ASR::abiType::BindC) {
+            if (all_symbols_mangling || should_mangle) {
+                ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+                sym_to_renamed[sym] = update_name(x.m_name);
+            }
+        }
+        for (auto &a : x.m_symtab->get_scope()) {
+            this->visit_symbol(*a.second);
+        }
+    }
+
+    void visit_ClassProcedure(const ASR::ClassProcedure_t &x) {
+        if (x.m_abi != ASR::abiType::BindC) {
+            if (all_symbols_mangling || should_mangle) {
+                ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+                sym_to_renamed[sym] = update_name(x.m_name);
+            }
+        }
+    }
+
+    void visit_AssociateBlock(const ASR::AssociateBlock_t &x) {
+        if (all_symbols_mangling || should_mangle) {
+            ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+            sym_to_renamed[sym] = update_name(x.m_name);
+        }
+        for (auto &a : x.m_symtab->get_scope()) {
+            this->visit_symbol(*a.second);
+        }
+    }
+
+    void visit_Block(const ASR::Block_t &x) {
+        if (all_symbols_mangling || should_mangle) {
+            ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+            sym_to_renamed[sym] = update_name(x.m_name);
+        }
+        for (auto &a : x.m_symtab->get_scope()) {
+            this->visit_symbol(*a.second);
+        }
+    }
+
+    void visit_Requirement(const ASR::Requirement_t &x) {
+        if (all_symbols_mangling || should_mangle) {
+            ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+            sym_to_renamed[sym] = update_name(x.m_name);
+        }
+        for (auto &a : x.m_symtab->get_scope()) {
+            this->visit_symbol(*a.second);
+        }
+    }
+
+    void visit_Template(const ASR::Template_t &x) {
+        if (all_symbols_mangling || should_mangle) {
+            ASR::symbol_t *sym = ASR::down_cast<ASR::symbol_t>((ASR::asr_t*)&x);
+            sym_to_renamed[sym] = update_name(x.m_name);
+        }
+        for (auto &a : x.m_symtab->get_scope()) {
+            this->visit_symbol(*a.second);
+        }
+    }
+
+};
+
+
+class UniqueSymbolVisitor: public ASR::BaseWalkVisitor<UniqueSymbolVisitor> {
```

Comment:
Document here that this visitor does the renaming (and only renaming) using the `std::map`.

Reply:
Done